tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py

```diff
@@ -35,6 +44,139 @@
+        new_node_name (str): The name of the new node that has been created
+        old_osd_node_names (list): The name of the new node that has been added to osd nodes
+        old_osd_ids (list): List of the old osd ids
+
+    Raises:
+        AssertionError: If the node replacement verification steps failed.
+
+    """
+    min_osd_nodes = 3
+    num_of_old_osd_nodes = len(old_osd_node_names)
+    ocs_nodes = node.get_ocs_nodes()
+    num_of_old_ocs_nodes = len(ocs_nodes)
+
+    if num_of_old_osd_nodes <= min_osd_nodes:
+        logger.info(
+            f"We have {num_of_old_osd_nodes} osd nodes in the cluster - which is the minimum number "
+            f"of osd nodes. Wait for the new created worker node to appear in the osd nodes"
+        )
+        timeout = 1500
+        # In vSphere UPI platform, we are creating new node with same name as deleted
+        # node using terraform
+        if (
+            config.ENV_DATA["platform"].lower() == constants.VSPHERE_PLATFORM
+            and config.ENV_DATA["deployment_type"] == "upi"
+        ):
+            new_osd_node_name = old_node_name
+        else:
+            new_osd_node_name = node.wait_for_new_osd_node(old_osd_node_names, timeout)
+        logger.info(f"Newly created OSD name: {new_osd_node_name}")
+        assert new_osd_node_name, (
+            f"New osd node not found after the node replacement process "
+            f"while waiting for {timeout} seconds"
+        )
+    elif num_of_old_osd_nodes < num_of_old_ocs_nodes:
+        num_of_extra_old_ocs_nodes = num_of_old_ocs_nodes - num_of_old_osd_nodes
+        logger.info(
+            f"We have {num_of_extra_old_ocs_nodes} existing extra OCS worker nodes in the cluster"
+            f"Wait for one of the existing OCS nodes to appear in the osd nodes"
+        )
+        timeout = 600
+        new_osd_node_name = node.wait_for_new_osd_node(old_osd_node_names, timeout)
+        assert new_osd_node_name, (
+            f"New osd node not found after the node replacement process "
+            f"while waiting for {timeout} seconds"
+        )
+    else:
+        logger.info(
+            f"We have more than {min_osd_nodes} osd nodes in the cluster, and also we don't have "
+            f"an existing extra OCS worker nodes in the cluster. Don't wait for the new osd node"
+        )
+        new_osd_node_name = None
+
+    assert node.node_replacement_verification_steps_ceph_side(
+        old_node_name, new_node_name, new_osd_node_name
+    )
+    assert node.node_replacement_verification_steps_user_side(
+        old_node_name, new_node_name, new_osd_node_name, old_osd_ids
+    )
+
+
+def delete_and_create_osd_node(osd_node_name):
+    """
+    Delete an osd node, and create a new one to replace it
+
+    Args:
+        osd_node_name (str): The osd node name to delete
+
+    """
+    new_node_name = None
+    old_osd_ids = node.get_node_osd_ids(osd_node_name)
```

Comment:
you are duplicating the node and device replacement func, we cannot reuse the existing func?